cocotb/types/array.py

```diff
@@ -0,0 +1,233 @@
+from typing import Optional, Any, Iterable, Iterator, overload
+from collections.abc import Sequence
+from .range import Range
+from functools import cached_property
+from itertools import zip_longest
+from sys import maxsize
+
+
+class Array(Sequence):
+    r"""
+    Fixed-size, arbitrarily-indexed, heterogeneous sequence type.
+
+    Arrays are similar to, but different from Python :class:`list`\ s.
+    An array can store values of any type or values of multiple types at a time, just like a :class:`list`.
+    Unlike :class:`list`\ s, an array's size cannot change.
+
+    The indexes of an array can start or end at any integer value, they are not limited to 0-based indexing.
+    Indexing schemes can be either ascending or descending in value.
+    An array's indexes are described using a :class:`~cocotb.types.Range` object.
+    Initial values are treated as iterables, which are copied into an internal buffer.
+
+    .. code-block:: python3
+
+        >>> Array("1234")  # the 0-based range `(0, len(value)-1)` is inferred
+        Array(['1', '2', '3', '4'], Range(0, 3))
+
+        >>> Array(range=Range(0, 'downto', -3))  # the initial values are `None`
+        Array([None, None, None, None], Range(0, -3))
+
+        >>> Array([1, True, object(), 'example'], Range(-2, 1))  # initial value and range lengths must be equal
+        Array([1, True, <object object at 0x7f4cff7b5570>, 'example'], Range(-2, 'to', 1))
+
+    Arrays also support "null" ranges; "null" arrays have zero length and cannot be indexed.
+
+    .. code-block:: python3
+
+        >>> Array(range=Range(0, 'to', 1))
+        Array([], Range(1, 'to', 0))
+
+    Indexing and slicing is very similar to :class:`list`\ s, except you use the indexing scheme you specified.
```

Comment:
```suggestion
    Indexing and slicing is very similar to :class:`list`\ s, except you use the indexing scheme specified.
```
?

Reply:
I might try to strike `you` from the documentation entirely, there are only ~3 uses.